jetty-client/src/main/java/org/eclipse/jetty/client/util/AbstractRequestContent.java

```diff
@@ -0,0 +1,250 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.client.util;
+
+import java.io.EOFException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractRequestContent implements Request.Content
+{
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractRequestContent.class);
+
+    private final AutoLock lock = new AutoLock();
+    private final String contentType;
+    private Subscription subscription;
+    private Throwable failure;
+
+    protected AbstractRequestContent(String contentType)
+    {
+        this.contentType = contentType;
+    }
+
+    @Override
+    public String getContentType()
+    {
+        return contentType;
+    }
+
+    @Override
+    public Subscription subscribe(Consumer consumer, boolean emitInitialContent)
+    {
+        Subscription oldSubscription;
+        Subscription newSubscription;
+        try (AutoLock ignored = lock.lock())
+        {
+            if (subscription != null && !isReproducible())
+                throw new IllegalStateException("Multiple subscriptions not supported on " + this);
+            oldSubscription = subscription;
+            newSubscription = subscription = newSubscription(consumer, emitInitialContent, failure);
+        }
+        if (oldSubscription != null)
+            oldSubscription.fail(new EOFException("Content replay"));
+        if (LOG.isDebugEnabled())
+            LOG.debug("Content subscription for {}: {}", this, consumer);
+        return newSubscription;
+    }
+
+    protected abstract Subscription newSubscription(Consumer consumer, boolean emitInitialContent, Throwable failure);
+
+    @Override
+    public void fail(Throwable failure)
```

Comment:
why is failure delivered to the content and not directly to the subscription?
If a specific request fails, that should only affect that requests subscription and not subsequent or concurrent subscriptions for the same content.

Reply:
> why is failure delivered to the content and not directly to the subscription?

Because the content may not have a subscription yet when the failure happens.
For example, you send a request with content, but all connections are busy and it remains queued (so there is no subscription yet). Time passes and the request fails because of total timeout before it has the chance to even be sent.

That is also why `newSubscription(...)` takes a failure argument: if the implementation subscribes exactly a moment after the content has been failed, then also the subscription is failed immediately.

Reply:
But that assumes that a Content can only be used for one request and there is nothing in the contract that makes that apparent.
I think failures have to be directly delivered to all Subscriptions.    IF failures are to be delivered to Content as well, that should be separate and perhaps only to non reproducible contents... or maybe only non reproducible contents act on them  or perhaps only a OneTimeContent sub interface has the failure callback?